a3p-integration/proposals/z:acceptance/test-lib/sync-tools.js

```diff
@@ -0,0 +1,285 @@
+ *  - operation: query dest account's balance
+ *  - condition: dest account has a balance >= sent token
+ * - Making sure an offer resulted successfully
+ *
+ */
+
+/**
+ * @typedef {object} RetyrOptions
+ * @property {number} maxRetries
+ * @property {number} retryIntervalMs
+ * @property {(...arg0: string[]) => void} log
+ * @property {(object) => void} [setTimeout]
+ * @property {string} [errorMessage=Error]
+ *
+ *
+ * @typedef {object} CosmosBalanceThresold
+ * @property {string} denom
+ * @property {number} value
+ */
+
+const ambientSetTimeout = globalThis.setTimeout;
+
+/**
+ * From https://github.com/Agoric/agoric-sdk/blob/442f07c8f0af03281b52b90e90c27131eef6f331/multichain-testing/tools/sleep.ts#L10
+ *
+ * @param {number} ms
+ * @param {*} sleepOptions
+ * @returns
+ */
+const sleep = (ms, { log = () => {}, setTimeout = ambientSetTimeout }) =>
+  new Promise(resolve => {
+    // @ts-ignore
+    log(`Sleeping for ${ms}ms...`);
+    setTimeout(resolve, ms);
+  });
+
+/**
+ * From https://github.com/Agoric/agoric-sdk/blob/442f07c8f0af03281b52b90e90c27131eef6f331/multichain-testing/tools/sleep.ts#L24
+ *
+ * @param {() => Promise} operation
+ * @param {(result: any) => boolean} condition
+ * @param {string} message
+ * @param {RetyrOptions} options
+ * @returns
+ */
+const retryUntilCondition = async (
+  operation,
+  condition,
+  message,
+  { maxRetries = 6, retryIntervalMs = 3500, log, setTimeout },
+) => {
+  console.log({ maxRetries, retryIntervalMs, message });
+  let retries = 0;
+
+  while (retries < maxRetries) {
+    try {
+      const result = await operation();
+      log('RESULT', result);
+      if (condition(result)) {
+        return result;
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        log(`Error: ${error.message}`);
+      } else {
+        log(`Unknown error: ${String(error)}`);
+      }
+    }
+
+    retries++;
+    console.log(
+      `Retry ${retries}/${maxRetries} - Waiting for ${retryIntervalMs}ms for ${message}...`,
+    );
+    await sleep(retryIntervalMs, { log, setTimeout });
+  }
+
+  throw Error(`${message} condition failed after ${maxRetries} retries.`);
+};
+
+/**
+ * Making sure a core-eval resulted succesfully deploying a contract
+ */
+const makeGetInstances = follow => async () => {
+  const instanceEntries = await follow(
+    '-lF',
+    `:published.agoricNames.instance`,
+  );
+
+  return Object.fromEntries(instanceEntries);
+};
+
+/**
+ *
+ * @param {string} contractName
+ * @param {{follow: () => object, setTimeout: (object) => void}} ambientAuthroity
+ * @param {RetyrOptions} options
+ * @returns
+ */
+export const waitUntilContractDeplyed = (
+  contractName,
+  ambientAuthroity,
+  options,
+) => {
+  const { follow, setTimeout } = ambientAuthroity;
+  const getInstances = makeGetInstances(follow);
+  const {
+    maxRetries = 6,
+    retryIntervalMs = 3500,
+    log = console.log,
+    errorMessage = 'Error',
+  } = options;
```

Comment:
This expression appears several times here. Can we extract a common function? My best guess would be this, but maybe you can come up with something better:
```suggestion
  const { maxRetries, retryIntervalMs, log, errorMessage } =
    overrideDefaultOptions(options);
```
Where `overrideDefaultOptions` is the only place that has to know the default values.

Reply:
resolved in https://github.com/Agoric/agoric-sdk/pull/10171/commits/57e25e6118e1925f9dcff04147c0bc928257ef36